### PR TITLE
Add `getVisitorKeys`

### PR DESCRIPTION
**Pull request by fisker**
## Checklist

- [ ] I’ve added tests to confirm my change works.
- [ ] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [ ] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [ ] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by fisker** [comment]
@thorn0 I added `getVisitorKeys` to the JS printer.

Before I go further, can you help me benchmark `b2e500b` against `c03a2cb`(next), I don't have a suitable environment to run the benchmark script. 🙏

**Comment by thorn0** [comment]
I'll run it in two hours when I get home.

**Comment by thorn0** [comment]
Okay, so what do I benchmark? The last commit here against the last commit in next?

**Comment by fisker** [comment]
`b2e500b` against `b2e500b` is enough, if not too trouble, "last commit here against the last commit in next" can be useful too.

**Comment by fisker** [comment]
To clarify, this work here is not to prove this is faster than #13236, even it result faster, add more keys in #13236 can get similar result.

Just try to implement a way I prefer, and try to not get it too slow.

**Comment by thorn0** [comment]
```
Benchmark 1: node ./bench.mjs b2e500b serial
  Time (mean ± σ):     45.203 s ±  0.897 s    [User: 59.402 s, System: 3.386 s]
  Range (min … max):   44.415 s … 46.862 s    10 runs

Benchmark 2: node ./bench.mjs b2e500b parallel
  Time (mean ± σ):     45.426 s ±  0.297 s    [User: 58.500 s, System: 5.061 s]
  Range (min … max):   44.902 s … 45.981 s    10 runs

Benchmark 3: node ./bench.mjs c03a2cb serial
  Time (mean ± σ):     47.870 s ±  0.109 s    [User: 61.952 s, System: 3.224 s]
  Range (min … max):   47.685 s … 48.002 s    10 runs

Benchmark 4: node ./bench.mjs c03a2cb parallel
  Time (mean ± σ):     49.170 s ±  0.412 s    [User: 62.448 s, System: 4.835 s]
  Range (min … max):   48.474 s … 49.765 s    10 runs

Summary
  'node ./bench.mjs b2e500b serial' ran
    1.00 ± 0.02 times faster than 'node ./bench.mjs b2e500b parallel'
    1.06 ± 0.02 times faster than 'node ./bench.mjs c03a2cb serial'
    1.09 ± 0.02 times faster than 'node ./bench.mjs c03a2cb parallel'
```

**Comment by fisker** [comment]
Not too bad, huh. Are you fine with this approach?

**Comment by thorn0** [comment]
I'd be okay if it were a bit faster.

**Comment by fisker** [comment]
@thorn0  Can you help me benchmark the last commit against next branch? I used it in comment attach, maybe improved.

**Comment by fisker** [comment]
It seems there are bugs in `getSortedChildNodes` the result not correctly sorted, can't do more improve, will look into it later. Won't include in this PR.

**Comment by fisker** [comment]
This last commit https://github.com/prettier/prettier/pull/13255/commits/47a093524f6f8e7cfc237abcb055220835357486 may be slower, but the code is much shorter, can you run benchmark on it too?

**Comment by thorn0** [comment]
```
Benchmark 1: node ./bench.mjs next serial
  Time (mean ± σ):     50.231 s ±  1.066 s    [User: 64.998 s, System: 3.444 s]
  Range (min … max):   48.290 s … 51.302 s    10 runs

Benchmark 2: node ./bench.mjs next parallel
  Time (mean ± σ):     49.017 s ±  0.551 s    [User: 62.331 s, System: 5.205 s]
  Range (min … max):   48.081 s … 49.792 s    10 runs

Benchmark 3: node ./bench.mjs 65e5b0a85 serial
  Time (mean ± σ):     46.191 s ±  1.366 s    [User: 60.448 s, System: 3.516 s]
  Range (min … max):   44.200 s … 47.606 s    10 runs

Benchmark 4: node ./bench.mjs 65e5b0a85 parallel
  Time (mean ± σ):     47.162 s ±  0.430 s    [User: 60.365 s, System: 5.089 s]
  Range (min … max):   46.543 s … 48.024 s    10 runs

Summary
  'node ./bench.mjs 65e5b0a85 serial' ran
    1.02 ± 0.03 times faster than 'node ./bench.mjs 65e5b0a85 parallel'
    1.06 ± 0.03 times faster than 'node ./bench.mjs next parallel'
    1.09 ± 0.04 times faster than 'node ./bench.mjs next serial'
```

**Comment by thorn0** [comment]
Benchmark 1: node ./bench.mjs next serial
  Time (mean ± σ):     48.598 s ±  0.145 s    [User: 62.672 s, System: 3.485 s]
  Range (min … max):   48.293 s … 48.768 s    10 runs

Benchmark 2: node ./bench.mjs next parallel
  Time (mean ± σ):     48.915 s ±  0.518 s    [User: 61.753 s, System: 4.899 s]
  Range (min … max):   48.265 s … 49.727 s    10 runs

Benchmark 3: node ./bench.mjs f116a4d serial
  Time (mean ± σ):     43.992 s ±  0.165 s    [User: 57.792 s, System: 3.494 s]
  Range (min … max):   43.636 s … 44.138 s    10 runs

Benchmark 4: node ./bench.mjs f116a4d parallel
  Time (mean ± σ):     44.501 s ±  0.254 s    [User: 57.442 s, System: 4.939 s]
  Range (min … max):   44.246 s … 45.141 s    10 runs

Summary
  'node ./bench.mjs f116a4d serial' ran
    1.01 ± 0.01 times faster than 'node ./bench.mjs f116a4d parallel'
    1.10 ± 0.01 times faster than 'node ./bench.mjs next serial'
    1.11 ± 0.01 times faster than 'node ./bench.mjs next parallel'

**Comment by fisker** [comment]
Wow, so  f116a4d little faster?

**Comment by thorn0** [comment]
I'll run these benchmarks again today, but i might not have time to review and play with the code. One thing I'm thinking about is that if the visitor keys are already in the right order, we might not need to sort children. Sounds risky, but potentially it's another good performance boost.

**Comment by fisker** [comment]
Added `to-fast-properties`, and tested that bundled code still work, hope it can be faster. 

- [Fast properties in V8](https://v8.dev/blog/fast-properties)

**Comment by thorn0** [comment]
> I thought about this too. ~But normally there are only a few(maybe 0-4) nodes, won't cost much to sort.~ Sorry, wrong. array may contain many nodes, but I don't think I want do that in this PR.

It also can't be done for all nodes, at least for Flow. Check [this AST for an object type](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEMCeAHOACAhtgXm2GwCMlswAabAbQBMBdCuGgMwoHNsBfAbgA6UEFRAQMMAJbQAzslC4AToogB3AApKEclLgA2q3GjmjSi3GADWcGAGVcAWzgAZSVDjI2+maxBmL1nYYFm6cyDCKAK6+cA6kcPT0Cc64UJyRuJxwAGIQig64MFJpyCC4kTAQIiAAFjAOegDqNZLwMsFgcLbarZIAbq1opWAyJiBuPoow6uacBZ7evgBWMgAetqF6cACKkRDwC3o+osGKk6VsemrVGIpuMI2S9DA1yAAcAAwnKj6N5hilW5wSZ9DyiACOe3gM3EOjKMgAtO4Eglqoo4JDJOiZpl5kgvEdfD4HJJwlEiZsdlCPPjFqIYLhSI9nq8kAAmenmSR6UIAYQgDjxIGBAFZqpEfAAVRk6AnHEB9aIASSgSVgtjAdwkAEFVbZ0FtDj4eDwgA). Maybe there exist other similar cases.

**Comment by thorn0** [comment]
```
Benchmark 1: node ./bench.mjs next serial
  Time (mean ± σ):     48.427 s ±  0.141 s    [User: 62.898 s, System: 3.541 s]
  Range (min … max):   48.194 s … 48.602 s    10 runs

Benchmark 2: node ./bench.mjs next parallel
  Time (mean ± σ):     48.645 s ±  0.299 s    [User: 61.481 s, System: 4.990 s]
  Range (min … max):   48.022 s … 49.091 s    10 runs

Benchmark 3: node ./bench.mjs f116a4d serial
  Time (mean ± σ):     44.006 s ±  0.142 s    [User: 58.062 s, System: 3.326 s]
  Range (min … max):   43.730 s … 44.224 s    10 runs

Benchmark 4: node ./bench.mjs f116a4d parallel
  Time (mean ± σ):     44.622 s ±  0.379 s    [User: 57.514 s, System: 5.174 s]
  Range (min … max):   44.090 s … 45.361 s    10 runs

Benchmark 5: node ./bench.mjs 65e5b0a85 serial
  Time (mean ± σ):     43.789 s ±  0.166 s    [User: 58.038 s, System: 3.315 s]
  Range (min … max):   43.463 s … 43.978 s    10 runs

Benchmark 6: node ./bench.mjs 65e5b0a85 parallel
  Time (mean ± σ):     44.617 s ±  0.392 s    [User: 57.861 s, System: 5.098 s]
  Range (min … max):   44.162 s … 45.388 s    10 runs

Summary
  'node ./bench.mjs 65e5b0a85 serial' ran
    1.00 ± 0.01 times faster than 'node ./bench.mjs f116a4d serial'
    1.02 ± 0.01 times faster than 'node ./bench.mjs 65e5b0a85 parallel'
    1.02 ± 0.01 times faster than 'node ./bench.mjs f116a4d parallel'
    1.11 ± 0.01 times faster than 'node ./bench.mjs next serial'
    1.11 ± 0.01 times faster than 'node ./bench.mjs next parallel'
```

**Comment by fisker** [comment]
Test the "Fast Prototype"? https://github.com/prettier/prettier/pull/13255/commits/b4df2d23609bade3171cbb0acd8ba27940f85a1f

**Comment by fisker** [comment]
Are you fine with the result? Should I continue?

**Comment by thorn0** [comment]
Yes, looks good.

**Comment by thorn0** [comment]
```
Benchmark 1: node ./bench.mjs next serial
  Time (mean ± σ):     48.502 s ±  0.208 s    [User: 63.117 s, System: 3.459 s]
  Range (min … max):   48.236 s … 48.893 s    10 runs

Benchmark 2: node ./bench.mjs next parallel
  Time (mean ± σ):     48.716 s ±  0.400 s    [User: 62.011 s, System: 5.092 s]
  Range (min … max):   48.253 s … 49.662 s    10 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 3: node ./bench.mjs f116a4d serial
  Time (mean ± σ):     43.995 s ±  0.159 s    [User: 58.019 s, System: 3.450 s]
  Range (min … max):   43.766 s … 44.264 s    10 runs

Benchmark 4: node ./bench.mjs f116a4d parallel
  Time (mean ± σ):     44.656 s ±  0.378 s    [User: 57.841 s, System: 4.973 s]
  Range (min … max):   44.184 s … 45.430 s    10 runs

Benchmark 5: node ./bench.mjs 65e5b0a85 serial
  Time (mean ± σ):     43.837 s ±  0.169 s    [User: 57.489 s, System: 3.403 s]
  Range (min … max):   43.494 s … 44.034 s    10 runs

Benchmark 6: node ./bench.mjs 65e5b0a85 parallel
  Time (mean ± σ):     44.621 s ±  0.370 s    [User: 58.116 s, System: 4.912 s]
  Range (min … max):   44.179 s … 45.289 s    10 runs

Benchmark 7: node ./bench.mjs b4df2d2 serial
  Time (mean ± σ):     43.993 s ±  0.243 s    [User: 58.062 s, System: 3.481 s]
  Range (min … max):   43.631 s … 44.395 s    10 runs

Benchmark 8: node ./bench.mjs b4df2d2 parallel
  Time (mean ± σ):     44.750 s ±  0.393 s    [User: 58.230 s, System: 5.162 s]
  Range (min … max):   44.358 s … 45.550 s    10 runs

Summary
  'node ./bench.mjs 65e5b0a85 serial' ran
    1.00 ± 0.01 times faster than 'node ./bench.mjs b4df2d2 serial'
    1.00 ± 0.01 times faster than 'node ./bench.mjs f116a4d serial'
    1.02 ± 0.01 times faster than 'node ./bench.mjs 65e5b0a85 parallel'
    1.02 ± 0.01 times faster than 'node ./bench.mjs f116a4d parallel'
    1.02 ± 0.01 times faster than 'node ./bench.mjs b4df2d2 parallel'
    1.11 ± 0.01 times faster than 'node ./bench.mjs next serial'
    1.11 ± 0.01 times faster than 'node ./bench.mjs next parallel'

**Comment by fisker** [comment]
TODO: `markdown` and `html`

File: src/main/comments.js [hunk 1]
```diff
@@ -18,66 +16,51 @@ import {
...
+      return [];
+    }
+
+    return canAttachComment(childNode)
+      ? childNode
+      : getSortedChildNodes(childNode, options);
+  });
+
+  // Sort by `start` location first, then `end` location
+  childNodes.sort(
```
Comment (thorn0) [inline src/main/comments.js:57]: We can't use `sort` here as it isn't stable and node can have the same loc data.

Comment (fisker) [reply src/main/comments.js:57]: Since v12, it's already stable https://v8.dev/features/stable-sort

Comment (thorn0) [reply src/main/comments.js:57]: As per the ES spec, it doesn't have to be stable, and Node isn't the only engine where Prettier is supposed to work.

Comment (thorn0) [reply src/main/comments.js:57]: Sorry, I was wrong. Didn't know that spec has changed.

File: src/language-js/traverse/visitor-keys.evaluate.js [hunk 2]
```diff
@@ -29,6 +29,8 @@ const additionalVisitorKeys = {
   BigIntTypeAnnotation: [],
   // This may invalid, need investigate
   TSAbstractMethodDefinition: ["decorators"],
+  // Babel missing this
+  Program: ["interpreter"],
```
Comment (thorn0) [inline src/language-js/traverse/visitor-keys.evaluate.js:33]: What is this property supposed to contain? A node?

Comment (fisker) [reply src/language-js/traverse/visitor-keys.evaluate.js:33]: Yes, a node called `InterpreterDirective`. https://github.com/fisker/prettier/blob/bf3ed96a84017fade16bc5e1e82bfe55a567b870/src/language-js/printer-estree.js#L207

https://github.com/fisker/prettier/blob/bf3ed96a84017fade16bc5e1e82bfe55a567b870/src/language-js/printer-estree.js#L785

**Comment by thorn0** [comment]
```
Benchmark 1: node ./bench.mjs next serial
  Time (mean ± σ):     51.899 s ±  0.404 s    [User: 66.620 s, System: 3.617 s]
  Range (min … max):   51.366 s … 52.545 s    10 runs

Benchmark 2: node ./bench.mjs next parallel
  Time (mean ± σ):     50.995 s ±  3.452 s    [User: 64.334 s, System: 5.256 s]
  Range (min … max):   48.232 s … 57.531 s    10 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Benchmark 3: node ./bench.mjs f116a4d serial
  Time (mean ± σ):     44.085 s ±  0.191 s    [User: 58.329 s, System: 3.205 s]
  Range (min … max):   43.718 s … 44.309 s    10 runs

Benchmark 4: node ./bench.mjs f116a4d parallel
  Time (mean ± σ):     44.901 s ±  0.432 s    [User: 58.259 s, System: 5.192 s]
  Range (min … max):   44.396 s … 45.645 s    10 runs

Benchmark 5: node ./bench.mjs 65e5b0a85 serial
  Time (mean ± σ):     43.961 s ±  0.179 s    [User: 58.193 s, System: 3.320 s]
  Range (min … max):   43.740 s … 44.237 s    10 runs

Benchmark 6: node ./bench.mjs 65e5b0a85 parallel
  Time (mean ± σ):     44.751 s ±  0.475 s    [User: 58.243 s, System: 5.106 s]
  Range (min … max):   44.184 s … 45.504 s    10 runs

Benchmark 7: node ./bench.mjs b4df2d2 serial
  Time (mean ± σ):     44.459 s ±  0.326 s    [User: 58.295 s, System: 3.372 s]
  Range (min … max):   44.066 s … 45.071 s    10 runs

Benchmark 8: node ./bench.mjs b4df2d2 parallel
  Time (mean ± σ):     45.369 s ±  0.270 s    [User: 58.603 s, System: 5.145 s]
  Range (min … max):   44.969 s … 45.718 s    10 runs

Benchmark 9: node ./bench.mjs f3aa9ccb9 serial
  Time (mean ± σ):     44.534 s ±  0.381 s    [User: 58.718 s, System: 3.327 s]
  Range (min … max):   43.907 s … 45.134 s    10 runs

Benchmark 10: node ./bench.mjs f3aa9ccb9 parallel
  Time (mean ± σ):     44.825 s ±  0.308 s    [User: 57.940 s, System: 5.011 s]
  Range (min … max):   44.466 s … 45.334 s    10 runs

Summary
  'node ./bench.mjs 65e5b0a85 serial' ran
    1.00 ± 0.01 times faster than 'node ./bench.mjs f116a4d serial'
    1.01 ± 0.01 times faster than 'node ./bench.mjs b4df2d2 serial'
    1.01 ± 0.01 times faster than 'node ./bench.mjs f3aa9ccb9 serial'
    1.02 ± 0.01 times faster than 'node ./bench.mjs 65e5b0a85 parallel'
    1.02 ± 0.01 times faster than 'node ./bench.mjs f3aa9ccb9 parallel'
    1.02 ± 0.01 times faster than 'node ./bench.mjs f116a4d parallel'
    1.03 ± 0.01 times faster than 'node ./bench.mjs b4df2d2 parallel'
    1.16 ± 0.08 times faster than 'node ./bench.mjs next parallel'
    1.18 ± 0.01 times faster than 'node ./bench.mjs next serial'

**Comment by thorn0** [comment]
It looks like benchmarking more than two commits at once leads to less accurate results.

File: src/language-css/clean.js [hunk 3]
```diff
@@ -9,6 +9,7 @@ const ignoredProperties = new Set([
   "before",
   "after",
   "trailingComma",
+  "spaces",
```
Comment (fisker) [inline src/language-css/clean.js:12]: This is an object, was treat as node before, and removed `.before`, `.after`.  It's not a node and only have `before`/`after`, we can just ignore it.

**Comment by fisker** [comment]
@thorn0 I think I've done here, hope you have time to review.

**Pull request merged**